lib/ddtrace/contrib/active_record/utils.rb

```diff
@@ -19,28 +21,25 @@ def self.adapter_port
           connection_config[:port]
         end
 
-        def self.connection_config(object_id = nil)
-          object_id.nil? ? default_connection_config : connection_config_by_id(object_id)
-        end
-
-        # Attempt to retrieve the connection from an object ID.
-        def self.connection_by_id(object_id)
-          return nil if object_id.nil?
-          ObjectSpace._id2ref(object_id)
-        rescue StandardError
-          nil
-        end
-
-        # Attempt to retrieve the connection config from an object ID.
-        # Typical of ActiveSupport::Notifications `sql.active_record`
-        def self.connection_config_by_id(object_id)
-          connection = connection_by_id(object_id)
-          return {} if connection.nil?
-
-          if connection.instance_variable_defined?(:@config)
-            connection.instance_variable_get(:@config)
+        # In newer Rails versions, the `payload` contains both the `connection` and its `object_id` named `connection_id`.
+        #
+        # So, if rails is recent we'll have a direct access to the connection.
+        # Else, we'll find it thanks to the passed `connection_id`.
+        #
+        # See this PR for more details: https://github.com/rails/rails/pull/34602
+        #
+        def self.connection_config(connection = nil, connection_id = nil)
+          if connection.nil? && connection_id.nil?
+            default_connection_config
           else
-            {}
+            conn =
+              connection || ::ActiveRecord::Base
+              .connection_handler
+              .connection_pool_list
```

Comment:
Will this iterate over all ActiveRecord connections? Regardless if they were made with `ActiveRecord::Base.establish_connection`, or from a class that inherits from `ActiveRecord::Base`? Just want to make sure connections configured in a legitimate way for Rails don't slip through the cracks.

Reply:
yep all there is connected (in whatever way) with AR

Reply:
I think, but I'm not sure, that it's all connections *in the current process*, because here's how `connection_pool_list` is defined (at least in Rails 4.2):

```ruby
      def connection_pool_list
        owner_to_pool.values.compact
      end
```

and here's how `owner_to_pool` is defined:

```ruby
      def owner_to_pool
        @owner_to_pool[Process.pid]
      end
```

If I understand well (and I'm absolutely not sure of that) this line of code: `@owner_to_pool[Process.pid]`, it says that it takes the pools in the current process only.

Which means that for a multi-processes runtime (like Puma with workers or Unicorn) the number of connections should be fairly low (at least, it's my assumption. I know that AR is by default configured to used 5 connections, IIRC)
And for a multi-threaded runtime (like Puma with JRuby) it means all the connections.

WDYT ?

Reply:
you got that right: one connection per thread 
per Model.establish_connection (`class_to_pool`) - which usually means a different DB connection.

`Process.pid` is really only due MRI - not much forking under JRuby.
under a multi-process (MRI) connections are low (usually ~1 since one process == one request), under a GIL-less runtime it can go as high as the server handles concurrent request processing.

Reply:
@kares Clever that since its segmented by process/thread that it makes the pool to be searched much smaller. I think this is only problematic if the AR event is generated from one thread, and handled in another, which would cause it to search the wrong pool. I can't really think of how that would happen here.

@guizmaii Okay, so I guess my last concern is, does this work with Rails 3.0 - latest? We have broken tests in this PR so its hard to tell. If we can get the build passing, it will run the Rails tests for these different versions, and that would be good enough to tell me whether this condition has been satisfied.